[dnf5] transaction history info and list

lukash opened the pull request:
I'm changing the loading of transaction history to be lazy, so that we can list all transactions when creating a query, but not necessarily load all the data.

There's a bit of an issue with the transactions from history being cached in the sack and the queries returning non-const objects meaning if someone ends up modifying one, the change will be present in the transaction cached in the sack.

Commit:
microdnf/commands/history: Add basic history list and history info

Commit:
libdnf/transaction/transaction: Move TransactionWeakPtr definition

Commit:
libdnf/transaction/transaction: Replace Sack and Query with TransactionHistory

Replaces the transaction::Sack and transaction::Query objects with a
single transaction::TransactionHistory object, which provides methods
for directly selecting transactions by ids (or all transactions).

The selected transactions are no longer cached, they are selected each
time and ownership is passed to the API user. This means the returned
objects rather should be copyable, hence the removal of all unique_ptrs
in the member variables throughout the hierarchy of classes.

j-mracek commented on include/libdnf-cli/output/key_value_table.hpp:
```diff
@@ -36,7 +36,6 @@ class KeyValueTable {
     ~KeyValueTable();
     void print();
 
-protected:
```
Do we really need it to make it public?

lukash replied:
Yes. It is the actual interface for using the class. When it was protected, the only way to use it was to inherit from the class.

I had no need for inheriting, I just wanted to use the class as it is (see the history info code).

kontura replied:
I am also doing this in the latest advisory PR for the very same reason.
So I would agree we want it public.

j-mracek replied:
Well advisory  can access it internally. It does not need to be as a part of API.

j-mracek replied:
Is it possible to create formatting functions like `print_transaction_table(Transaction & transaction)` that will use those functions internally?

lukash replied:
It makes no sense to try to hide this API, it is _the_ API that should be public for a libdnf-cli output class.

j-mracek commented on include/libdnf/transaction/transaction.hpp:
```diff
@@ -242,11 +241,25 @@ class Transaction {
+    /// Finish the transaction by updating it's state in the database
+    ///
+    /// @replaces libdnf:transaction/private/Transaction.hpp:method:Transaction.finish(libdnf::TransactionState state)
+    void finish(TransactionState state);
```
Why `start()` and `finish()` is public?

lukash replied:
They were public already, I just moved the code to have a nicer method order.

j-mracek replied:
Yes, but I think they should be private. Only DNF internels should allow to store transaction and anly transaction created internally without user modification + - elements. Only adding comments or transaction label should be allowed and may be with indirect modification.

m-blaha replied:
I tend to agree with @j-mracek here. I think `transaction::Transaction` objects should be read-only for API users. Is there a use case where creating / modifying Transaction objects via API is needed? History database records should be only created automatically during execution of `base.Transaction.run()`. It sounds weird  to me that API users can mess with the history database by creating new records there haphazardly.
Do you think this is a problem that deserves a TODO and a tracking ticket?

lukash replied:
I agree as well. It's on my list of things to do to finalize the API, I've made a note in the ticket we have for that and I've also added a TODO to a semi-random point in the code :slightly_smiling_face:

j-mracek commented on include/libdnf/transaction/transaction.hpp:
```diff
@@ -254,30 +255,30 @@ class Transaction {
 
-    std::set<std::string> runtime_packages;
-    std::vector<std::pair<int, std::string>> console_output;
+    mutable std::set<std::string> runtime_packages;
```
Why runtime_packages is a set? It somehow looks like an unused branch.

j-mracek replied:
I think it is a new thing in DNF5 - or I overlooked something in DNF4 DB scheme.

lukash replied:
I think the idea there is to record some select packages (of the dnf stack) with which the transaction was performed. I don't like it and I'm thinking about removing it. (if we want to record this, seems like serializing the packages we're interested in into a simple string and storing it in the comment or a similar single field is better than the complex structure that's used now)

I'm not sure if it's being filled up in dnf 4, but I was looking at the DB schema SQLs and they're identical between dnf 4 and dnf 5...

j-mracek replied:
I think the idea was nice but the detection is problematic. We can discuss it with the team.

Commit:
libdnf/transaction/db/rpm: Store integer into the rpm.epoch column

Surprisingly enough, sqlite allows storing any type of data into any
column. The code was storing an empty string into an integer column. It
seems it's a feature.

Commit:
libdnf/utils/sqlite3: Include SQL statement in SQLite3StatementSQLError

The SQL can be longish, but it's very important for debugging and just
logging it is a bit difficult because we don't pass Base to the SQLite
layer.
